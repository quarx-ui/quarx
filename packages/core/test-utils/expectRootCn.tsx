import React, { FC } from 'react';
import { render } from '@testing-library/react';

export const expectRootCn = <T extends FC<{ className?: string, children: any }>>(Component: T, name: string) => {
    const className = 'test-class-name';
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const { container: { firstChild: rootNode } } = render(<Component className={className} />);
    expect(rootNode).toHaveClass(className);
    expect(rootNode).toHaveClass(name);
};
