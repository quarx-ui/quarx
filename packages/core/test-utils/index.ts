export { expectRootCn } from './expectRootCn';
export * from './testBEM';
