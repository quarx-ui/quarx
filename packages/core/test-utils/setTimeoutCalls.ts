interface FakeSetTimeout {
    mock: {
        calls: Array<[() => void]>; // TODO Доработать типизацию
    };
}

/** При рендере jest использует setTimeout(_flushCallback, 0),
 * из-за чего проверка количества вызовов setTimeout происходит нестабильно.
 *
 * @return массив вызовов setTimeout без _flushCallback */
export const setTimeoutCalls = () => (setTimeout as unknown as FakeSetTimeout)
    .mock.calls
    .filter(([fn]) => !String(fn).includes('_flushCallback'));
