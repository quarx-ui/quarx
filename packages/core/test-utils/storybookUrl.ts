type PropsType = Record<string, string | boolean | number>;

interface UrlOptions {
    component: string;
    story?: string;
    props?: PropsType;
}

const reducePropsToURLString = (props?: PropsType) => {
    if (!props) {
        return '';
    }

    const keysOfProps = Object.keys(props);
    const valuesOfProps = Object.values(props);

    if (keysOfProps.length > 1) {
        return keysOfProps.reduce((prev, cur) => {
            const currentProp = `${cur}:${props[cur]}`;
            return `${prev};${currentProp}`;
        });
    }
    return `${keysOfProps[0]}:${valuesOfProps[0]}`;
};

export function storyUrl(options: UrlOptions) {
    const { component, story = 'sandbox', props } = options;
    const reducedProps = reducePropsToURLString(props);

    return `?path=/story/core-${component}--${story}&args=${reducedProps}`;
}
