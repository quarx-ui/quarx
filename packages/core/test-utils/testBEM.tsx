import React, { ComponentType } from 'react';
import { screen, render, cleanup } from '@testing-library/react';

export const expectPropsInClasses = (
    element: HTMLElement,
    qxClassname?: string,
) => (
    prop: string,
    value?: string | boolean,
) => {
    const qxSelector = qxClassname ?? element.classList[0];

    const valueIsBoolean = typeof value === 'boolean' || value === 'true' || value === 'false';
    const booleanClassname = value === true || value === 'true' ? `${qxSelector}_${prop}` : undefined;

    const expectedClassname = booleanClassname
        ?? (!valueIsBoolean
            ? `${qxSelector}_${prop}_${value}`
            : undefined);

    if (!expectedClassname) {
        return null;
    }

    return expect(element).toHaveClass(expectedClassname);
};

export const expectPropsMapInClasses = (element: HTMLElement, qxClassname?: string) => {
    const expectPropInClasses = expectPropsInClasses(element, qxClassname);

    return (expectedProps: Record<string, string | boolean>) => {
        Object.keys(expectedProps).forEach((prop) => {
            expectPropInClasses(prop, expectedProps[prop]);
        });
    };
};

export const testStyleParams = <
    S extends Record<string, any>,
    P extends Record<string, any> = Record<string, any>,
>(
    Component: ComponentType<P>,
    expectedDefaultParams: S,
    defaultProps: P = {} as P,
    qxClassname?: string,
) => (props: { [key in keyof S]: S[key][] }) => {
    describe('style params', () => {
        describe('default values', () => {
            const testId = 'testStyleParams-default';

            render(
                <Component
                    {...defaultProps}
                    data-testid={testId}
                />,
            );

            const element = screen.getByTestId(testId);

            Object.entries(expectedDefaultParams).forEach(([key, value]) => {
                it(`${key}`, () => {
                    expectPropsInClasses(element, qxClassname)(key, value);
                });
            });

            cleanup();
        });

        Object.entries(props).forEach(([propName, values]) => {
            describe(propName, () => {
                Object.values(values).forEach((value) => {
                    it(`${value}`, () => {
                        const testId = `testStyleParams-${propName}-${value}`;

                        const testProps = { [propName]: value };
                        const { asFragment } = render(
                            <Component
                                {...(defaultProps ?? {})}
                                {...testProps as P}
                                data-testid={testId}
                            />,
                        );

                        const element = screen.getByTestId(testId);
                        expectPropsMapInClasses(element, qxClassname)(testProps as S);
                        expect(asFragment()).toMatchSnapshot();
                    });
                });
            });
        });
    });
};
