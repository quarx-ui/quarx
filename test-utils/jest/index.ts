export { testRootCn } from './testRootCn';
export * from './testBEM';
