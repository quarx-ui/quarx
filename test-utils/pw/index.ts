export {
    compareStorySnapshots,
    compareVariantSnapshots,
    comparePropertySnapshots,
    comparePropertiesMapSnapshots,
} from './compareSnapshots';
