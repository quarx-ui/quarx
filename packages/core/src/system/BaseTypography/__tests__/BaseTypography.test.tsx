import { render, screen } from '@testing-library/react';
import {
    BASE_TYPOGRAPHY_TYPES,
    BaseTypography,
    BaseTypographyProps,
    BaseTypographyStyleParams,
    PALETTE_TEXT_KEYS,
    QX_SIZE, TYPOGRAPHY_WEIGHT,
} from '@core';
import { testStyleParams } from '@core/test-utils';

describe('BaseTypography', () => {
    testStyleParams<BaseTypographyStyleParams, BaseTypographyProps>(
        BaseTypography,
        {
            size: QX_SIZE.medium,
            type: BASE_TYPOGRAPHY_TYPES.text,
            color: PALETTE_TEXT_KEYS.main,
            fontWeight: TYPOGRAPHY_WEIGHT.normal,
        },
        {
            children: 'Текст для примера',
        },
    )({
        size: Object.values(QX_SIZE),
        type: Object.values(BASE_TYPOGRAPHY_TYPES),
        color: Object.values(PALETTE_TEXT_KEYS).concat(['customColor' as never]),
        fontWeight: Object.values(TYPOGRAPHY_WEIGHT),
    });
});

describe('BaseTypography behavior', () => {
    it('text should be in the document', () => {
        render(<BaseTypography>BaseTypography</BaseTypography>);

        expect(screen.queryByText('BaseTypography')).toBeInTheDocument();
    });
});
