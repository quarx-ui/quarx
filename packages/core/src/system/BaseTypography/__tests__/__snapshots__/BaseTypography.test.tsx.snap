// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTypography style params color constant 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_normal QxBaseTypography_color_constant css-bgxyyp-QxBaseTypography-root"
    data-testid="testStyleParams-color-constant"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params color customColor 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_normal QxBaseTypography_color_customColor css-uhehd2-QxBaseTypography-root"
    data-testid="testStyleParams-color-customColor"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params color main 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_normal QxBaseTypography_color_main css-18kc97n-QxBaseTypography-root"
    data-testid="testStyleParams-color-main"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params color secondary 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_normal QxBaseTypography_color_secondary css-cu0tcv-QxBaseTypography-root"
    data-testid="testStyleParams-color-secondary"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params color tertiary 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_normal QxBaseTypography_color_tertiary css-1mblp74-QxBaseTypography-root"
    data-testid="testStyleParams-color-tertiary"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params fontWeight black 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_black QxBaseTypography_color_main css-1wi2oin-QxBaseTypography-root"
    data-testid="testStyleParams-fontWeight-black"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params fontWeight bold 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_bold QxBaseTypography_color_main css-5r67ta-QxBaseTypography-root"
    data-testid="testStyleParams-fontWeight-bold"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params fontWeight extraBold 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_extraBold QxBaseTypography_color_main css-1b3dqq7-QxBaseTypography-root"
    data-testid="testStyleParams-fontWeight-extraBold"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params fontWeight extraLight 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_extraLight QxBaseTypography_color_main css-gs15me-QxBaseTypography-root"
    data-testid="testStyleParams-fontWeight-extraLight"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params fontWeight light 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_light QxBaseTypography_color_main css-1r0loxx-QxBaseTypography-root"
    data-testid="testStyleParams-fontWeight-light"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params fontWeight medium 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_medium QxBaseTypography_color_main css-4h6z42-QxBaseTypography-root"
    data-testid="testStyleParams-fontWeight-medium"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params fontWeight normal 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_normal QxBaseTypography_color_main css-18kc97n-QxBaseTypography-root"
    data-testid="testStyleParams-fontWeight-normal"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params fontWeight semibold 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_semibold QxBaseTypography_color_main css-1tbro2a-QxBaseTypography-root"
    data-testid="testStyleParams-fontWeight-semibold"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params fontWeight thin 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_thin QxBaseTypography_color_main css-u4i5v8-QxBaseTypography-root"
    data-testid="testStyleParams-fontWeight-thin"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params size large 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_large QxBaseTypography_fontWeight_normal QxBaseTypography_color_main css-yeg1ue-QxBaseTypography-root"
    data-testid="testStyleParams-size-large"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params size medium 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_normal QxBaseTypography_color_main css-18kc97n-QxBaseTypography-root"
    data-testid="testStyleParams-size-medium"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params size small 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_small QxBaseTypography_fontWeight_normal QxBaseTypography_color_main css-pfuynf-QxBaseTypography-root"
    data-testid="testStyleParams-size-small"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params size xLarge 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_xLarge QxBaseTypography_fontWeight_normal QxBaseTypography_color_main css-1yxhjv0-QxBaseTypography-root"
    data-testid="testStyleParams-size-xLarge"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params size xSmall 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_xSmall QxBaseTypography_fontWeight_normal QxBaseTypography_color_main css-1owll28-QxBaseTypography-root"
    data-testid="testStyleParams-size-xSmall"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params type headline 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_headline QxBaseTypography_size_medium QxBaseTypography_fontWeight_normal QxBaseTypography_color_main css-r5legd-QxBaseTypography-root"
    data-testid="testStyleParams-type-headline"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;

exports[`BaseTypography style params type text 1`] = `
<DocumentFragment>
  <div
    class="QxBaseTypography QxBaseTypography_type_text QxBaseTypography_size_medium QxBaseTypography_fontWeight_normal QxBaseTypography_color_main css-18kc97n-QxBaseTypography-root"
    data-testid="testStyleParams-type-text"
  >
    Текст для примера
  </div>
</DocumentFragment>
`;
